src/services/postPasteImportFixes.ts

```diff
@@ -0,0 +1,187 @@
+        if (imports.length > 0) {
+            insertImports(changes, targetFile, imports, /*blankLineBetween*/ true, preferences);
+        }
+        importAdder.writeFixes(changes, getQuotePreference(originalFile, preferences));
+        host.revertUpdatedFile?.(targetFile.fileName, updatedTargetFile.updatedFile.text, targetFile.text);
+    }
+    else {
+        const context: CodeFixContextBase = {
+            sourceFile: updatedTargetFile.updatedFile,
+            program: updatedTargetFile.originalProgram,
+            cancellationToken,
+            host,
+            preferences,
+            formatContext,
+        };
+        const importAdder = codefix.createImportAdder(updatedTargetFile.updatedFile, updatedTargetFile.updatedProgram, preferences, host);
+        forEachChild(updatedTargetFile.updatedFile, function cb(node) {
+            if (isIdentifier(node)) {
+                if (!updatedTargetFile.originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
+                    // generate imports
+                    importAdder.addImportsForUnknownSymbols(context, node, /*useAutoImportProvider*/ true);
+                }
+            }
+            else {
+                node.forEachChild(cb);
+            }
+        });
+        importAdder.writeFixes(changes, getQuotePreference(targetFile, preferences));
+        host.revertUpdatedFile?.(targetFile.fileName, updatedTargetFile.updatedFile.text, targetFile.text);
+    }
+    pastes.forEach(({ text, range }) => {
+        range.pos === range.end ? changes.replaceRangeWithText(targetFile, { pos: range.pos, end: range.end }, text) : changes.replaceRangeWithText(targetFile, { pos: range.pos, end: range.end }, text);
+    });
+}
+
+function getImportsFromKnownOriginalFile(
+    originalFile: SourceFile,
+    targetFile: SourceFile,
+    program: Program,
+    importAdder: codefix.ImportAdder,
+    importsToCopy: Map<Symbol, boolean>,
+    targetFileImportsFromOldFile: Set<Symbol>,
+    changes: textChanges.ChangeTracker,
+    preferences: UserPreferences,
+    host: LanguageServiceHost,
+    checker: TypeChecker,
+) {
+    const copiedOldImports: AnyImportOrRequireStatement[] = [];
+    importsToCopy.forEach((isValidTypeOnlyUseSite, symbol) => {
+        try {
+            importAdder.addImportFromExportedSymbol(skipAlias(symbol, checker), isValidTypeOnlyUseSite);
+        }
+        catch {
+            for (const oldStatement of originalFile.statements) {
+                forEachImportInStatement(oldStatement, i => {
+                    append(copiedOldImports, filterImport(i, factory.createStringLiteral(moduleSpecifierFromImport(i).text), name => importsToCopy.has(checker.getSymbolAtLocation(name)!)));
+                });
+            }
+        }

```

Comment:
I've moved this part of the code to utilities, but the `catch` part deals with unresolved imports. So, if an import statement is written for a file that doesn't exist, this part handles that case.

Reply:
The nesting still seems wrong. For every import that doesn’t resolve, it looks like you’re adding a syntactic copy of every import in the file. So if you had a file where 10 imports resolve and 2 don’t, you’d end up with 34 imports in the new file: 10 that were rewritten successfully via the ImportAdder; 12 when one fails; 12 again when the other fails. Maybe these are deduplicated down the line, but it at least appears that this is looping and synthesizing import nodes far too aggressively when things fail.

Reply:
Yeah, you're right I see how imports are getting duplicated here. I'll create an issue for this as a bug and fix it because I need to add more tests for it as well.